components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/graphQL/GraphQLQueryAnalysisHandler.java

```diff
@@ -0,0 +1,353 @@
+                    return -1;
+                }
+            } else {
+                return Collections.max(allocatedDepths);
+            }
+        } else {
+            return -1;
+        }
+    }
+
+    /**
+     * This method analyses the query
+     *
+     * @param messageContext message context of the request
+     * @param payload        payload of the request
+     * @return true, if the query is not blocked or false, if the query is blocked
+     */
+    private boolean analyseQuery(MessageContext messageContext, String payload) {
+        JSONParser jsonParser = new JSONParser();
+
+        try {
+            String graphQLAccessControlPolicy =
+                    (String) messageContext.getProperty(APIConstants.GRAPHQL_ACCESS_CONTROL_POLICY);
+            JSONObject policyDefinition = (JSONObject) jsonParser.parse(graphQLAccessControlPolicy);
+            if (analyseQueryDepth(messageContext, payload, policyDefinition) &&
+                    analyseQueryComplexity(messageContext, payload, policyDefinition)) {
+                return true;
+            } else {
+                return false;
+            }
+        } catch (ParseException e) {
+            log.error("Policy definition parsing failed. " + e.getMessage(), e);
+        }
+        return false;
+    }
+
+    /**
+     * This method analyses the query depth
+     *
+     * @param messageContext   message context of the request
+     * @param payload          payload of the request
+     * @param policyDefinition json object which contains the policy
+     * @return true, if the query depth does not exceed the maximum value or false, if query depth exceeds the maximum
+     */
+    private boolean analyseQueryDepth(MessageContext messageContext, String payload, JSONObject policyDefinition) {
+        String username = APISecurityUtils.getAuthenticationContext(messageContext).getUsername();
+
+        try {
+            String[] userRoles = getUserRoles(username);
+            int maxQueryDepth = getMaxQueryDepth(userRoles, policyDefinition);
+            if (maxQueryDepth > 0) {
+                MaxQueryDepthInstrumentation maxQueryDepthInstrumentation =
+                        new MaxQueryDepthInstrumentation(maxQueryDepth);
+                GraphQL runtime = GraphQL.newGraphQL(schema).instrumentation(maxQueryDepthInstrumentation).build();
+
+                try {
+                    ExecutionResult executionResult = runtime.execute(payload);
+                    List<GraphQLError> errors = executionResult.getErrors();
+                    if (errors.size() > 0) {
+                        List<String> errorList = new ArrayList<>();
+                        for (GraphQLError error : errors) {
+                            errorList.add(error.getMessage());
+                        }
+
+                        // TODO: https://github.com/wso2/carbon-apimgt/issues/8147
+                        ListIterator<String> iterator = errorList.listIterator();
```

Comment:
shall we replace this block with a lambda expression ?

Reply:
Not applicable anymore since the master branch has a different codebase.